runtime/lua/vim/shared.lua

```diff
@@ -190,4 +190,98 @@ function vim.pesc(s)
   return s:gsub('[%(%)%.%%%+%-%*%?%[%]%^%$]', '%%%1')
 end
 
+--- Type checking validation function
+---
+--- Examples:
+--- <pre>
+---  validate({ arg={ { 'foo' }, 'table' }})                          --> Nop
+---  validate({ arg={ 1, 'table' } })                                 --> error("'arg' is expected table, got number")
+---  validate({ arg1={ { 'foo' }, 'table' }, arg2={ 1, 'string' } })  --> error("'arg2' is expected string, got number")
+--- </pre>
+---
+---@param ... Table or list of table. That table is "argument_name = { validation_target, type_name (, whether nil is allowed) }".
+--- The following can be used as type_names:
+---     - table or tbl
+---     - string or str
+---     - number or num
+---     - boolean or bool
+---     - function or func
+---     - nil
+---     - thread
+---     - userdata
+function vim.validate(...)
+  local function _type_name(t)
```

Comment:
this function seems unnecessary. Maybe if the type aliases were 1 char I guess ('t' for table, 's', for string, ...?). But even then, it doesn't really improve readability or save much typing.